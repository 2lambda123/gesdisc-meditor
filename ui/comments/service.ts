import { validate } from 'jsonschema'
import type { User } from '../auth/types'
import { ErrorData } from '../declarations'
import { BadRequestException, UnauthorizedException } from '../utils/errors'
import { getCommentsDb } from './db'
import {
    CreateCommentUserInput,
    DocumentComment,
    UpdateCommentUserInput,
} from './types'
import {
    NewDocumentCommentUserInputSchema,
    UpdateDocumentCommentUserInputSchema,
} from './validation.schemas'

export async function createCommentAsUser(
    newComment: CreateCommentUserInput,
    user: User
): Promise<ErrorData<DocumentComment>> {
    try {
        const commentsDb = await getCommentsDb()

        if (!user?.uid) {
            throw new UnauthorizedException()
        }

        const validationResult = validate(
            newComment,
            NewDocumentCommentUserInputSchema
        )

        if (!validationResult.valid) {
            throw new BadRequestException(validationResult.toString())
        }

        const comment = await commentsDb.insertOne({
            ...newComment, // validated user input
            parentId: newComment.parentId || 'root', // TODO: Why not use undefined rather than 'root'? (refactor opportunity)
            userUid: user.uid,
            createdOn: new Date().toISOString(),
            createdBy: user.name,
            resolved: false, // can't create a resolved comment
        })

        return [null, comment]
    } catch (err: any) {
        console.error(err)

        return [err, null]
    }
}

export async function updateCommentAsUser(
    commentChanges: UpdateCommentUserInput,
    user: User
): Promise<ErrorData<DocumentComment>> {
    try {
        const commentsDb = await getCommentsDb()

        if (!user?.uid) {
            throw new UnauthorizedException()
        }

        const validationResult = validate(
            commentChanges,
            UpdateDocumentCommentUserInputSchema
        )

        if (!validationResult.valid) {
            throw new BadRequestException(validationResult.toString())
        }

        if (commentChanges.resolved) {
            // Resolving a comment is a special case since we need to resolve all the child comments as well.
            // Can safely return early after resolving as the validation rules ensure we aren't calling update to resolve while also updating other properties.
            const resolvedComment = await commentsDb.resolveComment(
                commentChanges._id,
                user.uid
            )

            return [null, resolvedComment]
        }

        const updatedComment = await commentsDb.updateCommentText(
            commentChanges._id,
            commentChanges.text
        )

        return [null, updatedComment]
    } catch (err: any) {
        console.error(err)

        return [err, null]
    }
}

export async function getCommentForDocument(
    commentId: string,
    documentTitle: string,
    modelName: string
): Promise<ErrorData<DocumentComment>> {
    try {
        const commentsDb = await getCommentsDb()

        const comment = await commentsDb.getCommentForDocument(
            commentId,
            documentTitle,
            modelName
        )

        return [null, comment]
    } catch (error) {
        console.error(error)

        return [error, null]
    }
}

export async function getCommentsForDocument(
    documentTitle: string,
    modelName: string
): Promise<ErrorData<DocumentComment[]>> {
    try {
        const commentsDb = await getCommentsDb()

        const comments = await commentsDb.getCommentsForDocument(
            documentTitle,
            modelName
        )

        return [null, comments]
    } catch (error) {
        console.error(error)

        return [error, null]
    }
}
