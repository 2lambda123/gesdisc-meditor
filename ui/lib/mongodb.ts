/**
 * MongoClient following the Vercel pattern
 * https://github.com/vercel/next.js/tree/canary/examples/with-mongodb
 */

import { MongoClient, MongoClientOptions } from 'mongodb'

const uri =
    process.env.MONGO_URL ||
    process.env.MONGOURL ||
    'mongodb://meditor_database:27017/'
const options: MongoClientOptions = {
    useUnifiedTopology: true,
}

let mongoClient: MongoClient
let mongoClientPromise: Promise<MongoClient>

if (process.env.NODE_ENV === 'development') {
    // In development mode, use a global variable so that the value
    // is preserved across module reloads caused by HMR (Hot Module Replacement).
    if (!globalThis._mongoClientPromise) {
        mongoClient = new MongoClient(uri, options)
        // @ts-ignore in development
        global._mongoClientPromise = mongoClient.connect()
    }
    // @ts-ignore in development
    mongoClientPromise = global._mongoClientPromise
} else {
    // In production mode, it's best to not use a global variable.
    mongoClient = new MongoClient(uri, options)
    mongoClientPromise = mongoClient.connect()
}

export const getDb = async (dbName?: string) => {
    return (await mongoClientPromise).db(dbName || process.env.DB_NAME)
}

// Export a module-scoped MongoClient promise. By doing this in a
// separate module, the client can be shared across functions.
export default mongoClientPromise
