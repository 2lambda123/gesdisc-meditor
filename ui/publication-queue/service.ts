import log from '../lib/log'
import natsClientPromise from '../lib/nats'
import type { QueueMessage } from './types'

const NATS_QUEUE_PREFIX = 'meditor-'

export function publishMessageToQueueChannel(
    channelName: string,
    message: QueueMessage
) {
    // we'll return a promise to indicate whether the publish succeeded or not
    return new Promise<string>(async (resolve, reject) => {
        const natsClient = await natsClientPromise
        const fullChannelName = !channelName.startsWith(NATS_QUEUE_PREFIX)
            ? NATS_QUEUE_PREFIX + channelName
            : channelName // ensure the channel name follows the "meditor-NAME" pattern

        log.info(`Publishing message to channel ${fullChannelName}`)
        log.debug(message)

        natsClient.publish(
            fullChannelName,
            JSON.stringify(message),
            (err: Error | undefined, guid: string) => {
                if (err) {
                    log.error('Failed to publish queue message ', err)
                    reject(err)
                    return
                }

                log.info('Successfully published queue message ', guid)
                resolve(guid)
            }
        )
    })
}
