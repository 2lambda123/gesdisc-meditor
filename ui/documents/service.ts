import Fuse from 'fuse.js'
import type { User, UserRole } from '../auth/types'
import type { ErrorData } from '../declarations'
import {
    constructEmailMessageForStateChange,
    shouldNotifyUsersOfStateChange,
} from '../email-notifications/service'
import { getModel, getModelWithWorkflow } from '../models/service'
import type { DocumentsSearchOptions, ModelWithWorkflow } from '../models/types'
import { publishMessageToQueueChannel } from '../publication-queue/service'
import { ErrorCode, HttpException } from '../utils/errors'
import { getTargetStatesFromWorkflow } from '../workflows/service'
import type { Workflow, WorkflowEdge, WorkflowNode } from '../workflows/types'
import { getDocumentsDb } from './db'
import type {
    Document,
    DocumentHistory,
    DocumentPublications,
    DocumentState,
} from './types'

const EMAIL_NOTIFICATIONS_QUEUE_CHANNEL =
    process.env.MEDITOR_NATS_NOTIFICATIONS_CHANNEL || 'meditor-notifications'
const DELETED_STATE = 'Deleted'

export async function getDocument(
    documentTitle: string,
    modelName: string,
    user: User,
    documentVersion?: string
): Promise<ErrorData<Document>> {
    try {
        const documentsDb = await getDocumentsDb()
        const userRolesForModel = findAllowedUserRolesForModel(modelName, user?.roles)

        const [modelError, { titleProperty = 'title', workflow }] =
            await getModelWithWorkflow(modelName)

        if (modelError) {
            throw modelError // failed to get the model associated with the document
        }

        const sourceToTargetStateMap = createSourceToTargetStateMap(
            userRolesForModel,
            workflow.edges
        )

        const document = await documentsDb.getDocument(
            documentTitle,
            documentVersion,
            modelName,
            sourceToTargetStateMap,
            titleProperty,
            user?.uid
        )

        if (!document) {
            throw new HttpException(
                ErrorCode.NotFound,
                `Requested document, ${documentTitle}, in model, ${modelName}, was not found`
            )
        }

        return [null, document]
    } catch (error) {
        console.error(error)

        return [error, null]
    }
}

// TODO: add OPTIONAL pagination (don't break existing scripts, perhaps the existence of pagination query params changes the output?)
export async function getDocumentsForModel(
    modelName: string,
    searchOptions?: DocumentsSearchOptions
): Promise<ErrorData<Document[]>> {
    try {
        const documentsDb = await getDocumentsDb()

        const [modelError, { titleProperty = 'title', workflow }] =
            await getModelWithWorkflow(modelName) // need the model to get the related workflow and title property

        if (modelError) {
            throw modelError
        }

        let documents = await documentsDb.getDocumentsForModel(
            modelName,
            searchOptions,
            titleProperty
        )

        if (searchOptions?.searchTerm) {
            // user is attempting a search. Mongo text search is VERY basic, so we'll utilize fuse.js to do the search
            const fuse = new Fuse(documents, {
                keys: [titleProperty], // TODO: investigate searching more than just the title property
            })

            // fuse.js returns search results with extra information, we just need the matching document
            documents = fuse
                .search(searchOptions.searchTerm)
                .map(searchResult => searchResult.item)
        }

        // add target states to documents
        documents = documents.map(document => ({
            ...document,
            'x-meditor': {
                ...document['x-meditor'],
                targetStates: getTargetStatesFromWorkflow(
                    document['x-meditor'].state,
                    workflow
                ), // populate document with states it can transition into
            },
        }))

        return [null, documents]
    } catch (error) {
        console.error(error)

        return [error, null]
    }
}

export async function getDocumentHistory(
    documentTitle: string,
    modelName: string
): Promise<ErrorData<DocumentHistory[]>> {
    try {
        const documentsDb = await getDocumentsDb()
        const [modelError, { titleProperty = '' }] = await getModel(modelName)

        if (modelError) {
            throw modelError
        }

        const historyItems = await documentsDb.getDocumentHistory(
            documentTitle,
            modelName,
            titleProperty
        )

        return [null, historyItems]
    } catch (error) {
        console.error(error)

        return [error, null]
    }
}

export async function getDocumentHistoryByVersion(
    versionId: string,
    documentTitle: string,
    modelName: string
): Promise<ErrorData<DocumentHistory>> {
    try {
        const documentsDb = await getDocumentsDb()
        const [modelError, { titleProperty = '' }] = await getModel(modelName)

        if (modelError) {
            throw modelError
        }

        const historyItem = await documentsDb.getDocumentHistoryByVersion(
            documentTitle,
            modelName,
            titleProperty,
            versionId
        )

        return [null, historyItem]
    } catch (error) {
        console.error(error)

        return [error, null]
    }
}

export async function getDocumentPublications(
    documentTitle: string,
    modelName: string
): Promise<ErrorData<DocumentPublications[]>> {
    try {
        const documentsDb = await getDocumentsDb()
        const [modelError, { titleProperty = '' }] = await getModel(modelName)

        if (modelError) {
            throw modelError
        }

        const publications = await documentsDb.getDocumentPublications(
            documentTitle,
            modelName,
            titleProperty
        )

        return [null, publications]
    } catch (error) {
        console.error(error)

        return [error, null]
    }
}

/**
 * one of the central parts of mEditor, responsible for transitioning a document through a workflow by changing it's state.
 *
 * after changing the document state, will notify the user and send out a publication message
 */
export async function changeDocumentState(
    documentTitle: string,
    modelName: string,
    newState: string, // must be a string, not enum, due to states not existing at compile time,
    user: User,

    // changeDocumentState options
    options?: {
        disableEmailNotifications?: boolean
        disableQueuePublication?: boolean
        dangerouslyUpdateDocumentProperties?: Document
    }
): Promise<ErrorData<Document>> {
    try {
        if (!newState) {
            throw new HttpException(ErrorCode.BadRequest, 'No state provided')
        }

        if (!user) {
            throw new HttpException(ErrorCode.Unauthorized, 'User is not logged in')
        }

        const documentsDb = await getDocumentsDb()
        const [modelError, model] = await getModelWithWorkflow(modelName)

        if (modelError) {
            // model must exist
            throw modelError
        }

        // fetch the requested document
        const [documentError, document] = await getDocument(
            documentTitle,
            modelName,
            user
        )

        if (documentError) {
            // document must exist
            throw documentError
        }

        // try to construct a new state, this will throw if any of the business rules fail
        const state = await constructNewDocumentState(document, model, newState, user)

        // got a new state, update the documents state in the database
        const ok = await documentsDb.addDocumentStateChange(
            document,
            state,
            options?.dangerouslyUpdateDocumentProperties
        )

        if (!ok) {
            // safety check, not sure how this would actually happen, but just in case it does, this stops the user from thinking the update went through
            //? why? because the underlying DB call would only fail if the document didn't exist. We just queried for it above and we never actually delete documents
            throw new HttpException(
                ErrorCode.InternalServerError,
                'Failed to change document state'
            )
        }

        // get the updated document from the database
        const [updatedDocumentError, updatedDocument] = await getDocument(
            documentTitle,
            modelName,
            user
        )

        if (updatedDocumentError) {
            throw updatedDocumentError
        }

        // send email notification of state change
        if (!options?.disableEmailNotifications) {
            await safelyNotifyOfStateChange(
                model,
                document,
                newState,
                getWorkflowEdgesMatchingSourceAndTarget(
                    model.workflow,
                    document['x-meditor'].state,
                    newState
                )[0],
                user
            )
        } else {
            console.debug(
                'User requested to change document state without sending email notifications'
            )
        }

        if (!options?.disableQueuePublication) {
            await safelyPublishStateChangeToQueue(model, document, newState)
        } else {
            console.debug(
                'User requested to change document state without publishing the state change to the queue'
            )
        }

        if (newState === DELETED_STATE) {
            await safelyDeleteDocument(model, document, user)
        }

        return [null, updatedDocument]
    } catch (error) {
        return [error, null]
    }
}

/**
 * responsible for constructing a new state for the document
 *
 * This has many business rules, the summarized version of the rules is:
 *
 *  - new state has to be a valid state in the workflow
 *  - new state must not be the same as the current state
 *  - user must be authenticated and have roles for the given model
 *  - workflow must be properly configured, no duplicate edges (how do we know which edge to follow to get to the requested state?)
 */
export async function constructNewDocumentState(
    document: Document,
    model: ModelWithWorkflow,
    newState: string,
    user: User
): Promise<DocumentState> {
    const targetStates = getTargetStatesFromWorkflow(
        document['x-meditor'].state,
        model.workflow
    )

    const matchingEdges = getWorkflowEdgesMatchingSourceAndTarget(
        model.workflow,
        document['x-meditor'].state,
        newState
    )

    //! can't transition to a state the document is already in
    if (newState === document['x-meditor'].state) {
        throw new HttpException(
            ErrorCode.BadRequest,
            `Cannot transition to state [${newState}] as the document is in this state already`
        )
    }

    //! can't transition to a state that isn't in the workflow
    if (targetStates.indexOf(newState) < 0) {
        throw new HttpException(
            ErrorCode.BadRequest,
            `Cannot transition to state [${newState}] as it is not a valid state in the workflow`
        )
    }

    //! can't transition to a state the user does not have permission to transition to
    if (document['x-meditor'].targetStates.indexOf(newState) < 0) {
        throw new HttpException(
            ErrorCode.BadRequest,
            `User does not have the permissions to transition to state ${newState}.`
        )
    }

    //! can't transition if the workflow has two edges with the same source and same target (how do we know which edge to follow?)
    if (matchingEdges.length !== 1) {
        throw new HttpException(
            ErrorCode.InternalServerError,
            `Workflow, ${model.workflow.name}, is misconfigured! There are duplicate edges from '${document['x-meditor'].state}' to '${newState}'.`
        )
    }

    // create the new document state!
    return {
        source: document['x-meditor'].state,
        target: newState,
        modifiedOn: new Date().toISOString(),
        modifiedBy: user.uid,
    }
}

export async function safelyNotifyOfStateChange(
    model: ModelWithWorkflow,
    document: Document,
    newState: string,
    currentEdge: WorkflowEdge,
    user: User
) {
    try {
        if (shouldNotifyUsersOfStateChange(newState, currentEdge)) {
            const emailMessage = await constructEmailMessageForStateChange(
                model,
                document,
                newState,
                currentEdge,
                user
            )

            // publish the email to the "notifications" queue channel
            //? A separate microservice, "meditor-notifier", is responsible for actually sending the email
            await publishMessageToQueueChannel(
                EMAIL_NOTIFICATIONS_QUEUE_CHANNEL,
                emailMessage
            )
        }
    } catch (err) {
        //! log the error but failing to send an email notification should NOT stop the state change as it is a side effect
        console.error(err)
    }
}

export async function safelyPublishStateChangeToQueue(
    model: ModelWithWorkflow,
    document: Document,
    state: string
) {
    try {
        if (shouldPublishStateChangeToQueue(model, state)) {
            // turns "Data Release" into "Data-Release"
            const channelName = model.name.replace(/ /g, '-')

            // publish the document state change to the right channel
            //? One or more subscribers can be subscribed to this particular channel, these are external subscribers
            //? (ex. an external subscriber that can publish "UMM-C" model documents to CMR)
            await publishMessageToQueueChannel(channelName, {
                id: document._id,
                document,
                model: {
                    titleProperty: model.titleProperty,
                },
                state,
                time: Date.now(),
            })
        }
    } catch (err) {
        //! log the error but failing to publish should NOT stop the state change as it is a side effect
        console.error(err)
    }
}

export async function safelyDeleteDocument(
    model: ModelWithWorkflow,
    document: Document,
    user: User
) {
    try {
        console.debug(
            `Handling delete document for ${model.name} - ${
                document[model.titleProperty]
            }`
        )

        const documentsDb = await getDocumentsDb()
        await documentsDb.deleteDocument(
            model,
            document[model.titleProperty],
            user.uid
        )

        console.debug(
            `Deleted ${model.name} - ${document[model.titleProperty]} (deleted by: ${
                user.uid
            })`
        )
    } catch (err) {
        //! log the error but don't block state change
        console.error(err)
    }
}

/**
 * should only block publishing a state change if the workflow is using "publishable" on states/nodes
 * and someone has explicitly set a node to not publishable
 */
export function shouldPublishStateChangeToQueue(
    model: ModelWithWorkflow,
    state: string
) {
    if (model.workflow.nodes.find(node => node.publishable)) {
        // this workflow supports "publishable"
        console.debug(
            `The workflow, ${model.workflow.name}, has at least one node with "publishable" set.`
        )

        const matchingNode = model.workflow.nodes.find(node => node.id === state)

        //! don't combine these into !matchingNode?.publishable, this is intentionally separately checking that the node exists AND is not publishable
        if (matchingNode && !matchingNode.publishable) {
            console.debug(
                `State, ${state}, is not publishable, skipping publication.`
            )
            return false
        }
    }

    // return true by default
    return true
}

/**
 * The workflow edges describe all possible transitions for a workflow. Given an array of roles, return as a map the workflow edges matching those roles.
 */
export function createSourceToTargetStateMap(
    userRolesForModel: string[] = [],
    workflowEdges: WorkflowEdge[] = []
): { [key: WorkflowEdge['source']]: WorkflowEdge['target'][] } {
    return workflowEdges.reduce((accumulator, edge) => {
        if (userRolesForModel.includes(edge.role)) {
            if (!Array.isArray(accumulator[edge.source])) {
                accumulator[edge.source] = []
            }

            accumulator[edge.source].push(edge.target)
        }

        return accumulator
    }, {})
}

/**
 * Given its inputs, this function returns an array of roles / actions a user has for a model. Users will have permissions defined in their roles property (see this function's typings). Where the model's name and the user's role.model name matches, a role (might be helpful to think of this as an actor, like "Author", "Reviewer", or "Publisher") will be returned for that model name.
 */
export function findAllowedUserRolesForModel(
    modelName: string = '',
    roles: User['roles'] = []
): UserRole['role'][] {
    return roles.reduce((accumulator, role) => {
        if (role.model === modelName) {
            return [...accumulator, role.role]
        }

        return accumulator
    }, [])
}

/**
 * Returns all workflow edges that a user's permissions allow.
 * todo: delete the following note after complete refactor of old mEditor API:
 * this is like `getDocumentModelMetadata`'s `that.sourceStates` and `that.targetStates` without actually finding unique source and target values (this gives you the base info to do that).
 */
function findAllowedWorkflowEdgesForUserRoles(
    roles: User['roles'] = [],
    workflow: Workflow
): WorkflowEdge[] {
    return workflow.edges.filter(workflowNode => {
        return !!roles.find(permission => permission.role === workflowNode.role)
    })
}

/**
 * Returns all workflow nodes that are marked `readyForUse: true`.
 * todo: delete the following note after complete refactor of old mEditor API:
 * this is `getDocumentModelMetadata`'s `that.readyNodes`
 */
function findWorkflowNodesReadyForUse(workflow: Workflow): WorkflowNode['id'][] {
    return workflow.nodes.reduce((accumulator, workflowNode) => {
        if (workflowNode.readyForUse) {
            return [...accumulator, workflowNode.id]
        }

        return accumulator
    }, [])
}

function getWorkflowEdgesMatchingSourceAndTarget(
    workflow: Workflow,
    source: string,
    target: string
): WorkflowEdge[] {
    return workflow.edges.filter(
        edge => edge.source === source && edge.target === target
    )
}
