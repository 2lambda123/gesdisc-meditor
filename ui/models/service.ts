import { BadRequestException, NotFoundException } from '../utils/errors'
import type { Model } from './types'
import jsonpath from 'jsonpath'
import type { ErrorData } from '../declarations'
import { getModelsDb } from './db'
import { getDocumentsDb } from '../documents/db'

type getModelOptions = {
    populateMacroTemplates?: boolean
    includeId?: boolean
}

/**
 * retrieves a model, optionally including macro templates
 *
 * Macro templates are a way to populate a model's schema at runtime with values from a different model
 *
 * For example, a model of "News" would have a field called "tags", tags is an enum.
 *
 * Instead of hardcoding tags in the "News" model, we set it to "enum": ["placeholder"] and replace that
 * with a list of all "Tags" model titles at runtime, so it becomes "enum": ["science", "space", ...]
 */
export async function getModel(
    modelName: string,
    options: getModelOptions = { includeId: true }
): Promise<ErrorData<Model>> {
    try {
        if (!modelName) {
            throw new BadRequestException('Model name is required')
        }

        const modelsDb = await getModelsDb()
        const model = await modelsDb.getModel(modelName)

        if (!model) {
            throw new NotFoundException(`Model not found: ${modelName}`)
        }

        // see top level documentation for description of macro templates
        if (options.populateMacroTemplates) {
            // validate the model's schema before continuing
            if (!this.isJson(model.schema)) {
                throw new BadRequestException(
                    `The schema for model, ${modelName}, contains invalid JSON`
                )
            }

            // execute the macro templates for this model and get their values
            // TODO: add support for macros
            //let populatedTemplates = await this.getPopulatedModelTemplates(model)
            let populatedTemplates = []

            // parse the schema into an object
            let schema =
                typeof model.schema === 'string'
                    ? JSON.parse(model.schema)
                    : model.schema

            // can also set macro templates for the layout, parse it's JSON as well if this model has a layout
            let layout = null

            if (model.layout && this.isJson(model.layout)) {
                layout =
                    typeof model.layout === 'string'
                        ? JSON.parse(model.layout)
                        : model.layout
            }

            // loop through each macro template and update any matching fields in the model
            populatedTemplates.forEach(template => {
                // update any jsonpath matches in the schema with the template values
                jsonpath.value(schema, template.jsonpath, template.result)

                // if model has a layout, check in the layout for any matching jsonpath to update
                if (layout && jsonpath.paths(layout, template.jsonpath).length) {
                    jsonpath.value(layout, template.jsonpath, template.result)
                }
            })

            // set the schema and layout back to JSON strings
            model.schema = JSON.stringify(schema, null, 2)

            if (layout) {
                model.layout = JSON.stringify(layout, null, 2)
            }
        }

        if (!options.includeId) {
            // drop the _id
            const { _id, ...modelWithoutId } = model
            return [null, modelWithoutId]
        }

        return [null, model]
    } catch (err) {
        return [err, null]
    }
}

export async function getModels(): Promise<ErrorData<Model[]>> {
    try {
        const modelsDb = await getModelsDb()
        const models = await modelsDb.getModels()

        return [null, models]
    } catch (err) {
        return [err, null]
    }
}

export async function getModelsWithDocumentCount(): Promise<ErrorData<Model[]>> {
    try {
        const [modelsError, models] = await getModels()

        if (modelsError) {
            throw modelsError
        }

        const documentsDb = await getDocumentsDb()

        // get a count of documents in each model
        const modelsWithDocumentCount = await Promise.all(
            models.map(async model => {
                const documentCount =
                    await documentsDb.getNumberOfUniqueDocumentsForModel(
                        model.name,
                        model.titleProperty
                    )

                return {
                    ...model,
                    'x-meditor': {
                        ...model['x-meditor'],
                        countAll: documentCount,

                        // this was originally a role-based count, but any role can see any document so it's always the same as countAll
                        // was able to greatly improve performance by just reusing the full count
                        count: documentCount,
                    },
                } as Model
            })
        )

        return [null, modelsWithDocumentCount]
    } catch (err) {
        return [err, null]
    }
}
