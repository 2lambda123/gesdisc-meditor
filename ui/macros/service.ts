import cloneDeep from 'lodash.clonedeep'
import { getModel } from '../models/service'
import type { Model, PopulatedTemplate, Template } from '../models/types'
import { onlyUnique } from '../utils/array'
import { ErrorCode, HttpException } from '../utils/errors'
import { getMacrosDb } from './db'

/**
 * Macros are a DSL for running queries used to populate models.
 * given a model with a property of templates, will populate the templates section with results
 *
 * Example, given a model with the property:
 * {
 *     templates: [
 *         "jsonpath": "$.properties.news.items.enum",
 *         "macro": "list News.title"
 *     ]
 * }
 *
 * Will return:
 * [
 *     "jsonpath": "$.properties.news.items.enum",
 *     "macro": "list News.title",
 *     "result": [
 *         ...a list of all News document titles
 *     ]
 * ]
 */
async function runModelTemplates(model: Model) {
    if (!model.hasOwnProperty('templates')) {
        return []
    }

    const templates: Model['templates'] = cloneDeep(model.templates)

    const populatedTemplates = await Promise.all(
        templates.map(async (template: Template) => {
            const [macroName, macroArgument] = template.macro.split(/\s+/)
            const macroService = macros.get(macroName)

            if (!macroService) {
                throw new HttpException(
                    ErrorCode.BadRequest,
                    `Macro, ${macroName}, not supported.`
                )
            }

            ;(template as PopulatedTemplate).result = await macroService(
                macroArgument
            )

            return template as PopulatedTemplate
        })
    )

    return populatedTemplates
}

// todo: same ErrorData tuple?
async function listUniqueFieldValues(macroArgument: string) {
    const macrosDb = await getMacrosDb()
    //* macroArgument looks like "Keywords.title" or "Collection%20Metadata.Combined_EntryID".
    const [encodedModelName, fieldName] = macroArgument.split('.')
    //* Model names stored as macros appear to be URL-component encoded. If that changes in the future, there is no harm in decoding an already-decoded string.
    const modelName = decodeURIComponent(encodedModelName)
    //! If the unwritten contract for the "list" macro is that the fieldName will always be the titleProperty, this is a useless argument to pass into getUniqueFieldValues.
    //* We need the title property: get the model described in the macro's argument without running the macro.
    const [error, { titleProperty }] = await getModel(modelName)

    if (error) {
        throw new HttpException(ErrorCode.NotFound, `Model not found: ${modelName}.`)
    }

    //* Do not await the result of this operation, as this is being consumed in Promise.all().
    const results = macrosDb.getUniqueFieldValues(fieldName, modelName, titleProperty)

    return results
}

/**
 * generates a JSONSchema dependencies section https://json-schema.org/understanding-json-schema/reference/conditionals.html
 *
 * The macro (defined in a model's templates section) would look something like this
 *
 *      jsonPath: $.dependencies
 *      macro: listDependenciesByTitle DAAC.cmrProviders[].CmrProvider
 *
 * The above macro would get a list of all DAACs with their cmrProviders and build a dependency tree out of it.
 * Selecting a DAAC would populate the "CmrProvider" field enum with a list of only that DAAC's CMR Providers.
 */
async function listDependenciesByTitle(dbo, item) {
    const [model, dependentField, targetField] = item[0].split('.')
    const modelName = decodeURIComponent(model)
    const mongoField = dependentField.replace(/\[\]$/, '') // remove brackets for the mongo query if this is an array field

    if (!model.match(/^\S+$/)) {
        console.log(
            "Error: collection name in '" + item + "' should not have white spaces"
        )
        throw 'Error: collection name in ' + item + ' should not have white spaces'
    }

    const results = await dbo
        .collection(modelName)
        .aggregate([
            { $sort: { 'x-meditor.modifiedOn': -1 } }, // Sort descending by version (date)
            { $group: { _id: '$title', doc: { $first: '$$ROOT' } } }, // Grab all fields in the most recent version
            { $replaceRoot: { newRoot: '$doc' } }, // Put all fields of the most recent doc back into root of the document
            {
                $project: {
                    _id: 0,
                    title: 1,
                    [mongoField]: 1,
                },
            },
        ])
        .toArray()

    // turn the list of results into a JSON Schema dependencies tree
    const dependencies = {
        [modelName]: {
            oneOf: results
                .filter(result => typeof result.title !== 'undefined') // filter out invalid results
                .map(result => ({
                    properties: {
                        [modelName]: {
                            enum: [result.title],
                        },
                        [targetField]: {
                            enum: result[mongoField],
                        },
                    },
                })),
        },
    }

    return dependencies
}

function userRoles(dbo) {
    var roleList = []
    return new Promise(function (resolve, reject) {
        dbo.collection('Models')
            .aggregate(
                [
                    {
                        $lookup: {
                            from: 'Workflows',
                            localField: 'workflow',
                            foreignField: 'name',
                            as: 'graph',
                        },
                    },
                    { $project: { _id: 0, name: 1, 'graph.roles': 1 } },
                    { $unwind: '$graph' },
                ],
                { allowDiskUse: true }
            )
            .toArray(function (err, res) {
                if (err) {
                    console.log(err)
                    throw err
                } else {
                    res.forEach(element => {
                        roleList.concat(element.graph.roles)
                    })
                    resolve(roleList.filter(onlyUnique))
                }
            })
    })
}

const macros = new Map<string, Function>()

macros.set('list', listUniqueFieldValues)

export { runModelTemplates }
