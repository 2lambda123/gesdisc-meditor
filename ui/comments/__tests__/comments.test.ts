import { Db } from 'mongodb'
import getDb from '../../lib/mongodb'
import alertsModel from '../../models/__test__/fixtures/models/alerts.json'
import { getCommentForDocument, getCommentsForDocument } from '../service'
import mockComments from './__fixtures__/comments.json'

const mockAlerts = [
    {
        title: 'Mock Alert without Comments',
    },
    { title: 'Mock Alert w/ Comments & Troublesome Title' },
]

describe('Comments Service', () => {
    let db: Db

    beforeAll(async () => {
        db = await getDb()

        await db.collection('Models').insertOne(alertsModel)
        await db.collection('Alerts').insertMany(mockAlerts)
        await db.collection('Comments').insertMany(mockComments)
    })

    afterAll(async () => {
        await db.collection('Models').deleteMany({})
        await db.collection('Alerts').deleteMany({})
        await db.collection('Comments').deleteMany({})
    })

    test('collection return all comments for an existing model and document with comments', async () => {
        const [error, comments] = await getCommentsForDocument({
            documentTitle: 'Mock Alert w/ Comments & Troublesome Title',
            modelName: 'Alerts',
        })

        expect(error).toBeNull()
        expect(comments).toHaveLength(3)
        expect(comments).toMatchSnapshot()
    })

    test('collection returns no comments for an existing model and document without comments', async () => {
        const [error, comments] = await getCommentsForDocument({
            documentTitle: 'Mock Alert without Comments',
            modelName: 'Alerts',
        })

        expect(error).toBeNull()
        expect(comments).toHaveLength(0)
        expect(comments).toMatchSnapshot()
    })

    // todo: figure out why this is failing; the underlying service works well when consumed through our API
    test.skip('singleton returns one comment for an existing model and document with comments', async () => {
        const [mockComment] = mockComments

        const [error, comment] = await getCommentForDocument({
            commentId: mockComment._id,
            documentTitle: mockComment.documentId,
            modelName: mockComment.model,
        })

        expect(error).toBeNull()
        expect(Object.keys(comment)).toHaveLength(11)
        expect(comment).toMatchSnapshot()
    })

    test('singleton returns no comment for an existing model and document without comments', async () => {
        const [error, comment] = await getCommentForDocument({
            commentId: '5c269eaa7f40f1002dfe85f1',
            documentTitle: 'Mock Alert w/ Comments & Troublesome Title',
            modelName: 'Alerts',
        })

        expect(error).toBeNull()
        expect(Object.keys(comment)).toHaveLength(0)
        expect(comment).toMatchSnapshot()
    })
})
