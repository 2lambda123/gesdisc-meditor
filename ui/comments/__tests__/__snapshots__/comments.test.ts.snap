// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments Service collection return all comments for an existing model and document with comments 1`] = `
Array [
  Object {
    "_id": "634de0e7e9cb15002e80a9c8",
    "createdBy": "Test User",
    "createdOn": "2022-10-04T14:52:31.105Z",
    "documentId": "Mock Alert w/ Comments & Troublesome Title",
    "lastEdited": "2022-10-04T19:27:18.020Z",
    "model": "Alerts",
    "parentId": "root",
    "resolved": true,
    "resolvedBy": "othertestuser",
    "text": "This is a comment, now with edits that improve the comment.",
    "userUid": "testuser",
    "version": "2022-08-15T15:17:04.357Z",
  },
  Object {
    "_id": "633de0e7e9ca15002e8089c6",
    "createdBy": "Test User",
    "createdOn": "2022-10-05T19:54:15.575Z",
    "documentId": "Mock Alert w/ Comments & Troublesome Title",
    "lastEdited": "2022-10-05T19:54:36.114Z",
    "model": "Alerts",
    "parentId": "root",
    "resolved": true,
    "resolvedBy": "othertestuser",
    "text": "This comment is will be resolved shortly. It also has edits!",
    "userUid": "testuser",
    "version": "2022-08-15T15:17:04.357Z",
  },
  Object {
    "_id": "634035c35e1978002fead5a1",
    "createdBy": "Test User",
    "createdOn": "2022-10-07T14:20:51.049Z",
    "documentId": "Mock Alert w/ Comments & Troublesome Title",
    "model": "Alerts",
    "parentId": "root",
    "resolved": false,
    "text": "Fresh comment that highlights an unresolved issue.",
    "userUid": "testuser",
    "version": "2022-08-15T15:17:04.357Z",
  },
]
`;

exports[`Comments Service collection returns no comments for an existing model and document without comments 1`] = `Array []`;

exports[`Comments Service singleton returns no comment for an existing model and document without comments 1`] = `Object {}`;

exports[`Comments Service singleton returns one comment for an existing model and document with comments 1`] = `Object {}`;
